New module to extend payroll

open-net-sarl opened the pull request:
```
* Add Worked Hours to contract.
* Add Hourly Rate to contract.
* Compute the Wage of contract, based on Worked Hours and Hourly Rate.
* Add Holiday Rate to contract.
* Add commision rate, commision and reimbursement to contract.
```

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Update metafiles

sbidoul committed:
[UPD] addons table in README.md

dcoopennet committed:
Add module hr_payroll_extended

dcoopennet committed:
Fix currency issue for reimbursement

dcoopennet committed:
Fix PEP8

dufresnedavid commented on hr_payroll_extended/models/hr_contract.py:
```diff
@@ -0,0 +1,98 @@
+        ExpensesObj = self.env['hr.expense']
+        for expense in ExpensesObj.search(filters):
+            sum += expense.account_move_id.amount
+        self.reimbursement += sum
```
sum is a very bad choice of variable in python

dufresnedavid commented on hr_payroll_extended/models/hr_contract.py:
```diff
@@ -0,0 +1,98 @@
+    commission = fields.Float(string='Commission',
+                              compute='_comp_wage_expenses')
+    comm_rate = fields.Float(string='Commissions Rate',
+                             digits=dp.get_precision('Payroll Rate'))
```
why not defining commission has a salary rule ?

dufresnedavid commented on hr_payroll_extended/models/hr_payslip.py:
```diff
@@ -0,0 +1,130 @@
+                if move_line_ids:
+                    move_line_ids = [str(id) for id in move_line_ids]
+                    q = """update account_move_line
+    set slip_id=%d where id in (%s)""" % (payslip.id, ','.join(move_line_ids))
```
You don't have to convert the ids as a string. Python does that out of the box.

... % (payslip.id, tuple(move_line_ids))

ghost commented:
The name of the module must be changed in order to be published on oca. It does not tell anything about what the module does. Most modules already in the repo extend the payroll in odoo.

elicoidal commented on hr_payroll_extended/README.rst:
```diff
@@ -0,0 +1,48 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Payroll Commissions and Reinbursements
```
@open-net-sarl  Thanks for your contribution to the OCA. 
I will perform a review on format and wording following this principles: 
- Contribution guidelines: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md
- Please follow the OCA README template:
  https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
- We recommend to include a screenshot as well to help new comers (nice to have)
- Please use simple headers in all py files following OCA templates
  https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

elicoidal commented on hr_payroll_extended/__init__.py:
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
```
Please use short headers.

elicoidal commented on hr_payroll_extended/data/hr.salary.rule.xml:
```diff
@@ -0,0 +1,29 @@
+        <field name="appears_on_payslip">True</field>
+        <field name="code">COMM</field>
+        <field name="condition_select">python</field>
+        <field name="name">Commissions sur prestations</field>
```
This is not a module for the Swiss localization anymore. Please use English for all rules names

elicoidal replied:
and codes

elicoidal commented on hr_payroll_extended/models/account_move_line.py:
```diff
@@ -0,0 +1,70 @@
+        if new_rec.account_id.internal_type == 'payable' and \
+           new_rec.partner_id:
+            query = """update account_move
+set partner_id=%d
```
improve indentation

elicoidal commented on hr_payroll_extended/models/hr_contract.py:
```diff
@@ -0,0 +1,98 @@
+from openerp import models, fields, api
+import openerp.addons.decimal_precision as dp
+
+import logging
```
remove those 2 lines

elicoidal commented on hr_payroll_extended/models/hr_contract.py:
```diff
@@ -0,0 +1,98 @@
+
+    @api.one
+    @api.depends('employee_id.user_id')
+    def _comp_wage_expenses(self):
```
CamelCase?

elicoidal commented on hr_payroll_extended/README.rst:
```diff
@@ -0,0 +1,48 @@
+    :alt: License: AGPL-3
+
+Payroll Commissions and Reinbursements
+=========================
```
Please include a line before and after the document title.
Line must be as long as the title

elicoidal commented on hr_payroll_extended/README.rst:
```diff
@@ -0,0 +1,48 @@
+    * Compute the Wage of contract, based on Worked Hours and Hourly Rate.
+    * Add Holiday Rate to contract.
+    * Add commision rate, commision and reimbursement to contract.
+
```
AFAICS there is more than that (link to expenses and invoices). Can you create a simple usage section and a use case?

elicoidal commented on hr_payroll_extended/README.rst:
```diff
@@ -0,0 +1,48 @@
+
+** Remarks: **
+    * As this module proposes its own report (same as the original, but with its own footer), don't forget to make it non-updatable.
+
```
Add a bug tracker section following standard README

elicoidal replied:
Add Try me on Runbot as well

elicoidal commented on hr_payroll_extended/README.rst:
```diff
@@ -0,0 +1,48 @@
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
https

elicoidal commented on hr_payroll_extended/__openerp__.py:
```diff
@@ -0,0 +1,55 @@
+
+
+{
+    'name': 'Extended - Payroll',
```
Name and module name are too generic IMO and mixes 2 concepts: extra fields and link with invoices/AML/expenses.
I would rather split it in 2 modules: hr_payroll_accounting and hr_payroll_commissions (or something similar)

elicoidal commented on hr_payroll_extended/__openerp__.py:
```diff
@@ -0,0 +1,55 @@
+        'hr_attendance',
+        'account'
+    ],
+    'version': '9.0.1.0',
```
9.0.1.0.0

elicoidal commented on hr_payroll_extended/__openerp__.py:
```diff
@@ -0,0 +1,55 @@
+        'hr_payroll_account',
+        'hr_contract',
+        'hr_attendance',
+        'account'
```
IMO no need to explicitly refer to account, hr_payroll, decimal_precision and hr_contract

elicoidal commented on hr_payroll_extended/__openerp__.py:
```diff
@@ -0,0 +1,55 @@
+    'name': 'Extended - Payroll',
+    'summary': 'Payroll Extended',
+    'category': 'Human Resources',
+    'author': "Open-Net Sàrl,Odoo Community Association (OCA)",
```
you might want to add utf support (u"something")?

elicoidal commented on hr_payroll_extended/__openerp__.py:
```diff
@@ -0,0 +1,55 @@
+    ],
+    'version': '9.0.1.0',
+    'auto_install': False,
+    'demo': [],
```
remove empty keys

elicoidal commented on hr_payroll_extended/models/account_move_line.py:
```diff
@@ -0,0 +1,70 @@
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    # ---------- Fields management
```
Remove those 2 comments. They dont add any information

dcoopennet committed:
Rename module

Fix PR issues

dcoopennet committed:
Flake8

dcoopennet committed:
Add images to manifest

dcoopennet committed:
Fix PEP8

dcoopennet committed:
Avoid bypassing ORM

dcoopennet committed:
Improved README

cyp-opennet commented:
:+1:

open-net-sarl committed:
Update hr.salary.rule.xml

dreispt commented on hr_payroll_commission/__openerp__.py:
```diff
@@ -0,0 +1,52 @@
+##############################################################################
+#
+#    OpenERP, Open Source Management Solution
+#    Copyright (C) 2004-TODAY OpenERP S.A. <http://www.openerp.com>
```
Was this originally authored by OpenERP SA?

dreispt replied:
Please remove OpenERP SA copyright if they are not authors.

dreispt commented on hr_payroll_commission/__openerp__.py:
```diff
@@ -0,0 +1,52 @@
+    'depends': [
+        'hr_payroll_account',
+        'hr_attendance',
+        'account'
```
I'm puzzled at why a Commissions feature would force me to install attendance and accounting.
Can't we isolate these dependencies?

dreispt commented on hr_payroll_commission/models/account_move_line.py:
```diff
@@ -0,0 +1,68 @@
+            account_move_obj = self.env['account.move']
+            move_ids = account_move_obj.search([('id', '=', rec.move_id.id)])
+            move_ids.write({'partner_id': rec.partner_id.id})
+        return ret
```
I would like to hear the @OCA/accounting-maintainers experts on this.

dreispt commented on hr_payroll_commission/models/hr_contract.py:
```diff
@@ -0,0 +1,95 @@
+
+    # ---------- Fields management
+
+    @api.one
```
deprecated

dreispt commented on hr_payroll_commission/models/invoices.py:
```diff
@@ -0,0 +1,38 @@
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    # ---------- Fields management
```
remove

dreispt commented on hr_payroll_commission/models/hr_contract.py:
```diff
@@ -0,0 +1,95 @@
+        self.commission += commission
+
+    worked_hours = fields.Float(string='Worked Hours')
+    hourly_rate = fields.Float(string='Hourly Rate')
```
The hourly rate is already provided by another module: https://github.com/OCA/hr/tree/9.0/hr_contract_hourly_rate

dreispt commented on hr_payroll_commission/README.rst:
```diff
@@ -0,0 +1,74 @@
+
+**For further information:**
+    * Commissions and reimbursements management: http://open-net.ch/blog/la-comptabilite-salariale-suisse-avec-odoo-1/post/salaires-avec-odoo-commissions-et-notes-de-frais-78
+
```
We should have a better explanation here.
In particular, I'm not familiar with the link between Commissions and Expenses. 
Maybe that's something specific to some countries.

I'm familiar with expense reimbursements paid on Payslips, but in my experience that has nothing to do with Commissions.

For this to be a universal feature, could you please expand a little on the use cases?

open-net-sarl replied:
I splitted these modules, there was no link between commissions and expenses reimbursement.

dcoopennet committed:
Splitter hr_payroll_commission in 2 modules : hr_payroll_commission and hr_payroll_expense

dcoopennet committed:
Set state to done on expense when payslip is confirmed

dcoopennet committed:
Change author in __openerp__.py

dreispt commented on hr_payroll_commission/README.rst:
```diff
@@ -0,0 +1,64 @@
+    * You also need to add a the rule COMM to the contract.
+    .. image:: http://s15.postimg.org/xa7roruij/Screen_Shot_02_12_16_at_02_58_PM.png
+    * When you have this properly setup, you just have to compute your payslip to find COMM into the list.
+    .. image:: http://s22.postimg.org/4ouwvh2f5/Screen_Shot_02_12_16_at_03_04_PM.png
```
I see also changes in Invoices on this module.
Can you explain them how Invoices are expected to relate to Commissions?

dreispt commented on hr_payroll_expense/README.rst:
```diff
@@ -0,0 +1,59 @@
+    * Add expenses reimbursement to Payslip.
+
+**For further information:**
+    *Expenses reimbursements management: http://open-net.ch/blog/la-comptabilite-salariale-suisse-avec-odoo-1/post/salaires-avec-odoo-commissions-et-notes-de-frais-78
```
An explanation on the feature is still needed here, even if a link for more info is provided.

open-net-sarl replied:
There is a usage example, isn't it enough ?

dreispt replied:
It can be, but it doesn't explain how Invoices relate to commissions.

open-net-sarl replied:
Are you speaking about hr_payroll_expense or hr_payroll_commission ?
Because in hr_payroll_commission there is a link between hr_payslip and account_invoice and maybe that's what you are talking about.

dcoopennet committed:
Fix pylint check isssues

dreispt commented:
@open-net-sarl Yes, the commissions module README does not explain how it makes use of Invoices.
Also, how does it work for people already using https://github.com/OCA/commission ?

open-net-sarl commented:
@dreispt The module hr_payroll_commission is not in conflict with commission repo module. But for people who are using this module are not gonna use the module hr_payroll_commission. These modules (hr_payroll_commission, hr_commission) implements a way to handle commission but it is not  the same approche.

open-net-sarl commented:
@dreispt 
We read the code of the other other module and the approach is quite different.
We are not based on sale, we are based on payment, and partial payment received in order to compute a bonus on the salary.

This is a need that is different on the need currently covered.

Both module could be even run in //, if you want to have a computation per sales ou an impact, cash basis, on the payroll

Thanks in advance

dcoopennet commented:
@dreispt @elicoidal Is there something i need to fix or any question ? If not please merge it.

dcoopennet commented:
@dreispt @elicoidal Can somebody review this PR and merge it if everything is OK now.
If not please tell us.

yvaucher commented:
@dreispt @elicoidal What is the status here ?

moylop260 commented on hr_payroll_commission/models/account_move_line.py:
```diff
@@ -0,0 +1,68 @@
+            if rec.account_id.internal_type != 'payable' or not rec.partner_id:
+                continue
+            account_move_obj = self.env['account.move']
+            move_ids = account_move_obj.search([('id', '=', rec.move_id.id)])
```
Why we need write `partner` in all lines?
What about use `account_move_line.partner_id`?

moylop260 replied:
What about where a multi-partner is required?
In the case of account.move created from hr.payroll there are many special cases.

elicoidal commented on hr_payroll_commission/README.rst:
```diff
@@ -0,0 +1,69 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Payroll Commissions
```
please add a line before the title

elicoidal commented on hr_payroll_commission/__init__.py:
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
```
please use standard simplified header of the OCA

elicoidal commented on hr_payroll_expense/README.rst:
```diff
@@ -0,0 +1,61 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Payroll Expenses Reinbursements
```
Please add a line before the title and use OCA README newest template

In general, please follow the principles: 
- Contribution guidelines: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md
- Please follow the OCA README template:
  https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
- We recommend to include a screenshot as well to help new comers (nice to have)
- Please use simple headers in all py files following OCA templates
  https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

dcoopennet committed:
Fixes on hr_payroll_commission
- Removing account_move_line.py (no more relevant)
- Fix headers
- Fix README.rst blank line

dcoopennet committed:
Fix hr_payroll_expense
- Fix headers
- Fix blank line on README.rst

dcoopennet committed:
Rebase README on the template for hr_payroll_expense

dcoopennet committed:
Rebase README on template for hr_payroll_comssion

dcoopennet committed:
Re-Adding account_move_line to add new field slip_id

dcoopennet committed:
Fix PEP8 new line

dufresnedavid commented on hr_payroll_expense/data/hr.salary.rule.xml:
```diff
@@ -0,0 +1,25 @@
+Created by dco@open-net.ch
+
+Copyright (c) 2014 Open-Net Ltd.
+-->
```
@elicoidal @dreispt I've been asking myself this question: Is it accepted by the community to put copyrights on data and views ? In my opinion, it seams a bit awkward.

elicoidal replied:
@dufresnedavid following https://github.com/OCA/maintainer-tools/blob/master/template/module/views/assets.xml and https://github.com/OCA/maintainer-tools/blob/master/template/module/data/model_name_data.xml, yes it is advised on data and views.
For data, I reckon that it is awkward sometimes for data.
To be noted that in this module, the copyright should be corrected to follow OCA templates

dufresnedavid commented on hr_payroll_expense/models/hr_contract.py:
```diff
@@ -0,0 +1,33 @@
+            reimbursement += expense.account_move_id.amount
+        self.reimbursement += reimbursement
+
+    reimbursement = fields.Float(string='Reimbursement',
```
fields declaration must be at the beggining of the class.

dcoopennet replied:
@dufresnedavid this field needs the method declared before.

dufresnedavid replied:
@dcoopennet nope, it does not

dcoopennet replied:
@dufresnedavid Maybe it's no more necessary now. But il thought it cannot call a method ("_comp_reimbursement") into the compute parameter of a field after the field declaration ?

dufresnedavid replied:
@dcoopennet the standard in OCA guidelines is to put the compute method after the field declaration. You only pass the method's name as a string. Odoo will figure how what method needs to be called. It does not need to be declared before.

dufresnedavid commented on hr_payroll_expense/models/hr_payslip.py:
```diff
@@ -0,0 +1,48 @@
+            ('slip_id', '=', self.id)
+        ])
+        for expense in expenses:
+            expenses.state = 'done'
```
You should call a method of hr.expense instead of directly assigning a value to the field state. What you are currently doing is manage the internals of a model from inside a method of another model. This is very bad practice.

dufresnedavid commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,87 @@
+        AccountMoveLineObj = self.env['account.move.line']
+        aml = AccountMoveLineObj.search([('slip_id', 'in', self.ids)])
+        if aml:
+            aml.write({'slip_id': False})
```
First, your method is too long. Seperate it in pieces. Second, what you are doing is very confusing. It seems a very bizare workflow.

dufresnedavid commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,87 @@
+                    move_line_ids = [str(id) for id in move_line_ids]
+                    q = """update account_move_line
+    set slip_id=%d where id in (%s)""" % (payslip.id, ','.tuple(move_line_ids))
+                    self.env.cr.execute(q)
```
First, you are injecting sql. Second, This looks like a hack. Why bypassing the orm ?

moylop260 replied:
@open-net-sarl @dcoopennet 

Please, Could you check the following guidelines?
[no-sql-injection](https://github.com/oca/maintainer-tools/blob/master/CONTRIBUTING.md#no-sql-injection)
[do-not-bypass-the-orm](https://github.com/oca/maintainer-tools/blob/master/CONTRIBUTING.md#do-not-bypass-the-orm)

ghost commented:
:-1: seriously, this module is confusing, no unit test, very bad programming

moylop260 commented:
@dufresnedavid excellent comments FYI We are working to identify sql-injection here:
https://github.com/OCA/pylint-odoo/pull/29
Your feedback is welcome!

dcoopennet committed:
Fixes asked by dufresnedavid
- Removing copyright on top of xml files
- Fields declaration moved to the top of the class
- Refactoring compute_sheet method in hr_payslip.py
- Calling method from hr_expense instead of changing state with write
- Removing SQL injection and ORM bybass into hr_payslip.py

dcoopennet committed:
Fix PEP8

moylop260 commented:
Could you add tests?
Could you check the following checks?

``` bash
************* Module hr_payroll_commission.models.hr_contract
hr_payroll_commission/models/hr_contract.py:21: [W8104(api-one-deprecated), HrContract._comp_commission] api.one deprecated
************* Module hr_payroll_expense
hr_payroll_expense/__init__.py:1: [W7908(missing-newline-extrafiles), ] views/hr_contract_view.xml Missing newline
hr_payroll_expense/__init__.py:1: [W7908(missing-newline-extrafiles), ] views/hr_payroll_view.xml Missing newline
************* Module hr_payroll_expense.models.hr_contract
hr_payroll_expense/models/hr_contract.py:16: [W8104(api-one-deprecated), HrContract._comp_reimbursement] api.one deprecated
```

feketemihai commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,72 @@
+
+    # ---------- Utilities
+
+    def _detach_invoices_from_payslip(self):
```
Why a new methos, isn't it easier to have self.invoices.unlink() in compute_sheet...

dcoopennet replied:
I don't want to delete the linked invoices, i just want to break the link between the 2 objects.

feketemihai replied:
Ok...thanks for explanation.

feketemihai commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,72 @@
+
+    # ---------- Fields management
+
+    invoices = fields.One2many('account.invoice', 'slip_id',
```
Please use invoice_ids for onemany...

feketemihai commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,72 @@
+
+    invoices = fields.One2many('account.invoice', 'slip_id',
+                               string='Invoices')
+    move_lines = fields.One2many('account.move.line', 'slip_id',
```
Please use move_line_ids for one2many

feketemihai commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,72 @@
+        if invoices:
+            invoices.write({'slip_id': False})
+
+    def _detach_move_lines_from_payslip(self):
```
The same as for invoices...unlink ...

feketemihai commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,72 @@
+        if aml:
+            aml.write({'slip_id': False})
+
+    def _attach_invoices_to_payslip(self):
```
This should be a get method that is depending on employee_id, and invoices field to have a compute method that should be launched at onchange employee...and no need to write on invoices, if you assign the invoice_ids field, it will already write on account.invoice model.

feketemihai commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,72 @@
+        invoice_ids.write({'slip_id': self.id})
+        return invoice_ids
+
+    def _attach_move_lines_to_payslip(self, invoices):
```
The same for the invoices...

feketemihai commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,72 @@
+            if not user_id:
+                continue
+
+            invoice_ids = self._attach_invoices_to_payslip()
```
What's the point of getting the invoices after you compute the payslip...if you already have a rule defined, no invoices linked the calculation will be zero...so first attach invoices and calculate payslip after

dreispt replied:
I concur that here is a logic duplication here.
IMO the right way is to first attach the move lines, and then compute the total sales value for commissions, and store it in a Payslip field (instead of a Contract field).
The above should be in a separate function, for extensibility.

feketemihai commented on hr_payroll_commission/models/hr_contract.py:
```diff
@@ -0,0 +1,38 @@
+        # Compute comission based
+        commission = 0
+        for invoice_id in invoice_ids:
+            for move_line in invoice_id.payment_move_line_ids:
```
How do you treat refund. if you reconcile a invoice with refund, the commision should be deducted, no?

feketemihai commented on hr_payroll_commission/models/hr_expenses.py:
```diff
@@ -0,0 +1,38 @@
+# -*- coding: utf-8 -*-
```
The file should be deleted, no, it's added in the hr_payroll_expense module

feketemihai commented on hr_payroll_expense/models/hr_payslip.py:
```diff
@@ -0,0 +1,47 @@
+                ('slip_id', '=', False),
+                ('state', '=', ['done', 'post']),
+            ]
+            expenses = ExpenseObj.search(filters)
```
The same as the commision module, the payslip is calculated and you add expenses after...

feketemihai commented on hr_payroll_commission/views/hr_employee_view.xml:
```diff
@@ -0,0 +1,25 @@
+      <field name="type">form</field>
+      <field name="arch" type="xml">
+        <field name="children" position="after">
+          <field name="children_student"/>
```
Where is this field defined...

dcoopennet committed:
Fixes suggested by fekemithai
- Change fields name
- Compute payslip after commission or expenses computation

dcoopennet committed:
Fix PEP8 and remove unused files

dcoopennet committed:
Adding unittest and fixing issues

dcoopennet committed:
Improve unitest on hr_payroll_commission

dcoopennet committed:
Fix PEP8

dcoopennet committed:
Add unittest on hr_payroll_expense

feketemihai commented on hr_payroll_commission/README.rst:
```diff
@@ -0,0 +1,90 @@
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues
```
@open-net-sarl Could you please change the project_repo variable to hr.

feketemihai commented on hr_payroll_commission/README.rst:
```diff
@@ -0,0 +1,90 @@
+<https://github.com/OCA/{project_repo}/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed feedback.
+
```
Could you also add Try me on runbot paragraph...

feketemihai commented on hr_payroll_commission/data/hr.salary.rule.xml:
```diff
@@ -0,0 +1,17 @@
+<odoo>
+<data>
+    <record id="hr_payroll.hr_rule_comm" model="hr.salary.rule">
```
The id should be without "hr_payroll."

feketemihai commented on hr_payroll_commission/data/hr.salary.rule.xml:
```diff
@@ -0,0 +1,17 @@
+        <field name="quantity">1</field>
+        <field name="condition_range">contract.wage</field>
+        <field name="sequence">300</field>
+        <field name="category_id" ref="hr_payroll.ALW"/>
```
The link to ALW category should be with your module name, not hr_payroll., or you want  to direct link commision with ALW salary rule defined in hr payroll?, if you want that remove your free alowance category defined upper...

feketemihai commented on hr_payroll_commission/models/account_move_line.py:
```diff
@@ -0,0 +1,11 @@
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    slip_id = fields.Many2one('hr.payslip', string='Pay slip')
```
Payslip, without space...

feketemihai commented on hr_payroll_commission/models/invoices.py:
```diff
@@ -0,0 +1,14 @@
+
+    # ---------- Fields management
+
+    slip_id = fields.Many2one('hr.payslip', string='Pay slip')
```
Payslip, without space..

feketemihai commented on hr_payroll_commission/tests/test_commission.py:
```diff
@@ -0,0 +1,130 @@
+        )
+        self.partner = self.env.ref("base.res_partner_2")
+        self.product = self.env.ref("product.product_product_4")
+        self.account_receivable = self.env['account.account'].search(
```
You can use direct self.env.ref, without the search...

dcoopennet replied:
self.env.ref returns in this case the user_type of the account not the account.
Is there a self.env.ref to get directely an account of this type ?

feketemihai replied:
I think you can directly use self.env.ref with l10n_generic_coa.conf_a_recv since at demo it will install the generic chart of accounts...

feketemihai commented on hr_payroll_expense/README.rst:
```diff
@@ -0,0 +1,82 @@
+===========
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/{project_repo}/issues>`_. In case of trouble, please
```
Could you change project_repo variable to hr...and add Try me on Runbot paragraph..

feketemihai commented on hr_payroll_commission/data/hr.salary.rule.category.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
```
File name on xml data should have "_" instead of ".", only csv files should be names with ".", and for all files from 2 modules..

feketemihai commented on hr_payroll_commission/models/invoices.py:
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
```
Please change name of file with account_invoice..

feketemihai commented on hr_payroll_expense/models/hr_payslip.py:
```diff
@@ -0,0 +1,46 @@
+class HrPayslip(models.Model):
+    _inherit = 'hr.payslip'
+
+    expenses = fields.One2many('hr.expense', 'slip_id',
```
expense_ids should be the field name

feketemihai commented on hr_payroll_expense/models/hr_payslip.py:
```diff
@@ -0,0 +1,46 @@
+        if expenses:
+            expenses.write({'slip_id': False})
+
+        res = super(HrPayslip, self).compute_sheet()
```
The same as upper, isn't super be launched after you compute expenses...

dcoopennet replied:
The super must be called before the slip id is set on the expenses (hr_payroll_expense) or on the invoices (hr_payroll_commission). It also must be called after the expenses or the invoices are detached from the employee's payslip. I know it's a be difficult to understand but it's a way to handle the case where 2 draft employee's payslip compute.

dreispt replied:
Isn't there a risk that the computed commission doesn't match the move lines attached?
Can you explain better the issue with the 2 draft payslips for the same employee? You mean two contracts, two periods?

dcoopennet replied:
I mean if your employee has 2 payslip in draft, that means the module needs to apply commission/expense only on one of these payslips. This is the main reason why these values commission/reimbursement are placed in the contract.

feketemihai commented on hr_payroll_commission/README.rst:
```diff
@@ -0,0 +1,90 @@
+Payroll Commissions
+===================
+
+This module allows you to add commission and reimbursement to payslip
```
remove reimbursement, and detail about commission being calculated on received payments..

dcoopennet committed:
Fix issues mentionned by feketemihai

yvaucher commented on hr_payroll_commission/views/hr_contract_view.xml:
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
When using `<odoo>` tag no need to use `<data>` tag

yvaucher commented on hr_payroll_expense/README.rst:
```diff
@@ -0,0 +1,85 @@
+   :alt: License: AGPL-3
+
+===============================
+Payroll Expenses Reinbursements
```
s/Reinbursements/Rei**m**bursements/

yvaucher commented on hr_payroll_expense/models/hr_expenses.py:
```diff
@@ -0,0 +1,13 @@
+class HrExpenseExpense(models.Model):
+    _inherit = 'hr.expense'
+
+    # ---------- Fields management
```
Line to remove

yvaucher commented on hr_payroll_expense/views/hr_contract_view.xml:
```diff
@@ -0,0 +1,28 @@
+            </field>
+        </record>
+    </data>
+</odoo>
```
Missing EOL at EOF

yvaucher commented on hr_payroll_expense/views/hr_payroll_view.xml:
```diff
@@ -0,0 +1,28 @@
+            </field>
+        </record>
+    </data>
+</odoo>
```
Another missing end of line

yvaucher commented on hr_payroll_commission/models/hr_contract.py:
```diff
@@ -0,0 +1,39 @@
+    @api.depends('employee_id.user_id')
+    def _comp_commission(self):
+        for contract in self:
+            contract.commission = 0
```
What the point of setting `contract.commission` to 0 to add it `commission` value later?

keeping only the line ``contract.commission = commission` should be enough

dreispt replied:
@open-net-sarl can you comment?

dcoopennet committed:
Fixes suggested by @yvaucher

dcoopennet commented:
Is there anything blocking this PR to be merged ?

feketemihai commented:
@OCA/human-resources-maintainers before merge we should do a fast forward of 8 branch in 9 since most of v8 modules are upgraded and ported to new api...

dreispt commented:
@feketemihai I didn't get that. AFAIK between branches we do ports, not fast-forwards.

dreispt commented on hr_payroll_commission/models/hr_contract.py:
```diff
@@ -0,0 +1,39 @@
+    _inherit = 'hr.contract'
+
+    # ---------- Fields management
+    commission = fields.Float(string='Commission',
```
This name is misleading. is is not the amount of commissions, but rather the amount of invoiced sales eligible for commission.

dreispt replied:
- [ ] Can we improve this field name?

dreispt commented on hr_payroll_commission/models/hr_payslip.py:
```diff
@@ -0,0 +1,76 @@
+        # self.invoice_ids.unlink()
+        # self.move_line_ids.unlink()
+        self._detach_invoices_from_payslip()
+        self._detach_move_lines_from_payslip()
```
I think the new API correct  way is:

```
for payslip in self:
    payslip.invoice_ids = None
    payslip.payment_move_line_ids = None
```

dreispt commented on hr_payroll_commission/README.rst:
```diff
@@ -0,0 +1,92 @@
+   :alt: License: AGPL-3
+
+===================
+Payroll Commissions
```
This module really computes commissions on invoiced sales.
The title and description should say that.

dreispt commented on hr_payroll_expense/models/hr_contract.py:
```diff
@@ -0,0 +1,33 @@
+    @api.depends('employee_id.user_id')
+    def _comp_reimbursement(self):
+        for contract in self:
+            contract.reimbursement = 0
```
Redundant, remove.

dreispt commented on hr_payroll_expense/models/hr_contract.py:
```diff
@@ -0,0 +1,33 @@
+
+            # Compute reimbursement
+            reimbursement = 0
+            ExpensesObj = self.env['hr.expense']
```
Move up, outside the main loop.

dreispt commented on hr_payroll_expense/models/hr_contract.py:
```diff
@@ -0,0 +1,33 @@
+    _inherit = 'hr.contract'
+
+    reimbursement = fields.Float(string='Reimbursement',
+                                 compute='_comp_reimbursement')
```
Can we place this field in the Payslip instead?

dcoopennet replied:
I don't think it's a good idea because an employee can have multiple payslip in draft at the same time.
If this was placed in payslip how can we choose the payslip where to put the reimbursement.

dreispt commented:
Reviewed: 
Functionally, the two unbundled feature now make sense to me.
Technically: I have concerns about the implementation in both modules, it needs continuing the discussion and improvement.
Should be good to merge once concerns are solved.

dcoopennet commented:
Thanks for your review @dreispt, don't hesitate to ask more questions.
I will commit soon a fix containing your suggestions.

dcoopennet committed:
Fix suggestions from @dreispt

dreispt commented:
@pedrobaeza closed this by accident?

pedrobaeza commented:
It has been auto closed when recreating the 9.0 branch as requested by @feketemihai. Let's see if reopen works

pedrobaeza commented:
A rebase is needed, please.

dreispt commented:
OK

dcoopennet committed:
Fix rebase issue

open-net-sarl commented:
Sorry but i leave this PR open one week and then i'll close it.
I will push theses modules on apps.odoo.com

dreispt commented:
@open-net-sarl This was in WIP status, waiting for a rebase. You didn't comment that, but I take that it's ready for review now - I'm removing the WIP tag!
